scrape/scrape.go

```diff
@@ -1808,12 +1810,16 @@ loop:
 			break loop
 		}
 
-		if !seriesCached {
+		// If series wasn't cached (is new, not seen on previous scrape) we need need to add it to the scrape cache.
+		// But we only do this for series that were appended to TSDB without errors.
+		// If a series was new but we didn't append it due to sample_limit or other errors then we don't need
+		// it in the scrape cache because we don't need to emit StaleNaNs for it when it disappears.
+		if !seriesCached && sampleAdded {
+			ce = sl.cache.addRef(met, ref, lset, hash)
```

Comment:
With the latest change check for `ref == 0` in `addRef()` is no longer needed - https://github.com/prometheus/prometheus/pull/16429/files#diff-4535d6172e0a9ab4e0879d71ef67ab69023f6eda9b44100343a6afa04fc8925fR1077